feat(listView): add ListView components

jtomasek opened the pull request:
<!--
Thanks for your interest in patternfly-react. We appreciate all issues filed and PRs submitted!

Please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
**What**:
This change builds up on the ListView solution introduced by https://github.com/patternfly/patternfly-react/pull/32, makes it a bit more flexible and adds a list of base components which can be used as building blocks to implement custom ListViewItem component

<!-- Why are these changes necessary? -->

<!-- How were these changes implemented? -->
**How**:
There are 3 main components:
ListView - a wrapper for ListViewItems
ListViewItem - handles the logic of expanding and general ListViewItem behavior
ListViewRow - Defines structure and contents of each ListViewItem

Note: ListView and ListViewItem could be used to implement List and ListItem components (https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-pf.html) in future as they basically implement the desired behavior. IMHO this matches the general PatternFly goal to merge these two content views.


<!-- feel free to add additional comments -->
Storybook available here: http://rawgit.com/jtomasek/patternfly-react/listView-storybook/index.html

priley86 commented:
hey @jtomasek, thanks a lot for providing this. I'll be diving in more this week to review it...

One thing I notice after a quick glance, would this component support the Component Expansion pattern [here] (https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-compound-expansion.html)? That seems to be the only pattern I don't see here in the stories...

Also - agree that once we have decided on the API for this one, the `List` and `ListItem` should match it as close as possible. Thanks again for submitting this!

jtomasek commented:
@priley86 I've looked into integrating compound expansion into ListViewItem and it turns out that it would make the ListViewItem implementation much more complicated as the toggling becomes much more complex. Also It makes the ListViewItem logic diverge from the ListItem.

Can we consider Compound expansion a subject for a custom ListViewItem implementation? I have updated the ListViewExpand component to support usable inside expandable ListViewInfoItem, so it should be possible to implement custom ListViewItem which supports the use case. I can work on creating a story for that, although it would be better if I could do it in a separate PR.

priley86 commented:
@jtomasek thanks for the detailed response ;). I don't think that it's a huge concern and am fine to treat this pattern separately (via custom ListViewItem). I just wanted to pose the question after quickly glancing...

@jgiardino any thoughts on the compound expansion pattern? Is this something we anticipate we'll continue supporting in the new List patterns?

jgiardino commented:
I've been out this week and am traveling today. I'd like some time to review the html in more detail. But regarding the compound expansion pattern, I have not heard that this is going away. However, I don't object to implementing this as a separate component.

waldenraines commented on src/ListView/ListView.stories.js:
```diff
@@ -0,0 +1,193 @@
+              <Col sm={11}>
+                {item.expandedContentText}
+              </Col>
+            </Row>
```
I like how the `ListViewItem` knows there is an expansion based solely on if there is a child node rather than requiring an explicit property.

waldenraines commented on src/ListView/ListViewItem.js:
```diff
@@ -0,0 +1,113 @@
+      children,
+      stacked,
+      actions,
+      additionalItems,
```
It seems that this is called both `additionalItems` as well as `ListViewAdditionalInfo`.  I wonder if we should converge on one name to avoid confusion.

jtomasek replied:
Would changing the prop to additionalInfo be sufficient?

priley86 commented on src/ListView/ListViewItem.js:
```diff
@@ -0,0 +1,113 @@
+export default class ListViewItem extends React.Component {
+  constructor() {
+    super()
+    this.state = { expanded: false }
```
question - do we prefer state object defined inside the constructor or outside within the class? We don't have a linter checking this right now, but it may be good to introduce it at some point for consistency... maybe we can consider more lint settings in a future PR?

jtomasek replied:
All examples in ReactJS doc initialize the state in constructor.

priley86 replied:
@jtomasek I'm good with this pattern (just wanted to note it though). Will update my other PRs accordingly and follow this standard.  Thanks!

jgiardino commented on src/ListView/ListViewRow.js:
```diff
@@ -0,0 +1,76 @@
+  heading,
+  description
+}) =>
+  <div style={{ display: 'flex', flex: 1 }}>
```
I’m curious why you included the inline styles here. The only difference I see in layout when these styles are applied vs removed is that the border between the checkbox and title is either the height of the `list-view-pf-main-info` or the height of the `list-view-pf-checkbox`. This is the only visual difference I can see with the core implementation. 

Did you have an issue with the layout when these styles were not included?

jtomasek replied:
Omitting the styling breaks the flexbox chain and causes the row not to span across whole width of parent. This effect is visible in storybook ListView > ListView variants - 3rd item - if you remove the styling there, it will shorten to the left, making the title very narrow.

This styling is needed because ListViewRow adds additional div. I know that bootstrap 4 has a special class for this, bootstrap 3 does not.

jgiardino replied:
Thanks for the clarification! I see what you're referring to now. My only suggestion would be to include these styles under a class, that way we can use this class in other components where similar issues occur. I'm fine with that being tracked as a separate issue outside this PR, though.

jgiardino commented:
Thanks for contributing this, @jtomasek! It's pretty amazing how you were able to include so much flexibility in the list view implementation.

There were just a couple of things I noticed:

- The CDN link to patternfly in storybook is not the latest version of patternfly. The styles for list view have been updated since the version being used in storybook (e.g. there's a box-shadow around the expanded list row). Could you update the version to be the latest?
- When the list row is expanded, the class `list-view-pf-expand-active` should be added to `div.list-group-item`
- Plus one inline comment regarding inline styles.

jtomasek commented:
Thanks all for a review!
Updates:
- added latest PatternFly from CDN
- renamed additionalItems prop to additionalInfo
- fixed active state of list-group-item

priley86 commented:
@jtomasek thanks again for all the hard work on this. After discussing w/ @jgiardino, we'd like to merge this PR and follow up on the inline styles/accessibility of this component in future PRs (also in our new implementation of `list-pf` #31).
